### ISSUE TITLE
Questions on Comment data

### ISSUE THREAD

**Issue by arvindshmicrosoft** [ldbc/ldbc_snb_datagen_spark#325]
Hello! I have 2 questions on the Comment data.

#### Q1: Why is ParentPostId not included when a Comment is a replyOf another Comment?
When a comment is a reply to another comment, the ParentCommentId is populated, but ParentPostId is null. It would make things a bit easier for partitioned data loads if the ParentPostId is also included for such cases (by making it a direct join back to the forum where the comment belongs, rather than a recursive query just for this data transformation). See for example the below, with the _italicized_ value indicating the proposed change:
| creationDate | id | locationIP | browserUsed | content | length | CreatorPersonId | LocationCountryId | ParentPostId | ParentCommentId |
|---|---|---|---|---|---|---|---|---|---|
| 2/3/2011 23:01 | 412316860804 | 196.29.42.107 | Firefox | About Franz Kafka, part of the English languageAbout Erwin Rommel,  it was his le | 81 | 42 | 60 | 412316860802 | |
| 3/10/2012 18:52 | 893353197958 | 192.160.111.235 | Internet Explorer | About Jamaica, s. Once a Spanish possessioAbout George IV of the United Kingdo | 78 | 910 | 49 | _412316860802_  | 412316860804 |
| 3/11/2012 5:07 | 893353197959 | 196.29.42.107 | Firefox | good | 4 | 42 | 60 | _412316860802_  | 893353197958 | 

#### Q2: Given a single output partition file (part*.csv), are the ParentCommentId values expected / guaranteed to be in the same file, for those cases when a Comment is in replyOf another Comment?
In other words, is a single comment hierarchy always scoped to within a single partition? I'm guessing this is the case, but wanted to check if this ensured by design and implementation.

**Comment by szarnyasg** [comment]
Hi @arvindshmicrosoft 

Regarding Q1:
> Why is ParentPostId not included when a Comment is a replyOf another Comment?

This is a deliberate choice – if systems are want to know the `ParentPostId`, they have to compute it by traversing the Message tree. This is true both during the bulk load phase and also during inserts (in the Interactive workload's small transactions and also in the larger - daily - batches of the BI workload).

I'll let David chime in regarding Q2.

**Comment by dszakallas** [comment]
> Given a single output partition file (part*.csv), are the ParentCommentId values expected / guaranteed to be in the same file, for those cases when a Comment is in replyOf another Comment?

We do not guarantee this. Even for the raw mode, where we have a custom serializer for writing the tree-based structure of posts and comments into tabular format. It uses a modulo based technique where the nth row of a Spark partition outputs  to  a file `part_<partition_num>_<file_num>` where `file_num` is `n mod m` (m is a constant specific to the entity type).  This is done in order to control individual file sizes. Another straightforward way is to split into fixed length chunks, but even that wouldn't get completely rid of this problem. Moreover, in BI mode inserts and deletes are repartitioned by their timestamp which can shuffle around comments even more.

**Comment by arvindshmicrosoft** [comment]
Thanks for the clarifications. I was looking at the [Neo4J trillion entity demo](https://github.com/neo4j/trillion-graph/blob/master/stores.md) and noticed that while it mentions LDBC, the output is quite customized. Specifically, that generator puts out shards based on the forumId.

So, I am curious if you would consider explicitly adding a forumId field in the official datagen to allow more efficient usage of sharding if needed by the implementing platform?

**Comment by szarnyasg** [comment]
> So, I am curious if you would consider explicitly adding a forumId field in the official datagen to allow more efficient usage of sharding if needed by the implementing platform?

It depends on how we define the "official datagen":
* For the Datagen configuration that should be used for benchmarking, I'd say *no* as such a flag may make the benchmark simpler for certain systems. That said, if an implementation would like to reorganize the data set (by e.g. adding redundant data such as a message's forum id or parent post id), it can do so during the loading phase. Then, the combined "reorganization plus loading" time should be disclosed in the benchmark report.
* As a configuration flag for the Datagen that can be used to generate data sets for e.g. internal testing, I'd be open to adding such a flag.

**Comment by arvindshmicrosoft** [comment]
Thank you @szarnyasg - I will close this Issue as we have had a good discussion above. I will let you folks consider the request to possibly add an extra flag to possible add the forumId as a test-only extra field to the related outputs.
